File: tests/selenium/catroid/DetailsTests.php
<?php
/*    Catroid: An on-device graphical programming language for Android devices
 *    Copyright (C) 2010-2011 The Catroid Team
 *    (<http://code.google.com/p/catroid/wiki/Credits>)
 *
 *    This program is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU Affero General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU Affero General Public License for more details.
 *
 *    You should have received a copy of the GNU Affero General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* Set TESTS_BASE_PATH in testsBootstrap.php to your catroid www-root */
require_once 'testsBootstrap.php';

class DetailsTests extends PHPUnit_Framework_TestCase
{
  private $selenium;

  public function setUp()
  {
    $path= 'http://'.str_replace('http://', '', TESTS_BASE_PATH).'catroid/';
    $this->selenium = new Testing_Selenium(TESTS_BROWSER, $path);
    if (TESTS_SLOW_MODE==TRUE) {
      $this->selenium->setSpeed(TESTS_SLOW_MODE_SPEED);
    } else {
      $this->selenium->setSpeed(1);
    }
    $this->selenium->start();
  }

  public function tearDown()
  {
    $this->selenium->stop();
  }

  public function ajaxWait()
  {
    for($second = 0; $second <= 600; $second++) {
      if($second >= 600) break;
      try {
        if($this->selenium->isElementPresent("xpath=//input[@id='ajax-loader'][@value='off']")) {
          break;
        }
      } catch (Exception $e) {}
      sleep(1);
    }
  }

  /**
   * @dataProvider randomIds
   */
  public function testDetailsPage($id, $title, $description)
  {
    $this->selenium->open(TESTS_BASE_PATH.'catroid/details/'.$id);
    $this->selenium->waitForPageToLoad(10000);

    //project title
    $this->assertEquals($title, $this->selenium->getText("xpath=//div[@class='detailsProjectTitle']"));

    //test the view counter
    $numOfViews = intval($this->selenium->getText("xpath=//p[@class='detailsStats']/b"));
    $this->selenium->refresh();
    $this->selenium->waitForPageToLoad(10000);
    $numOfViewsAfter = intval($this->selenium->getText("xpath=//p[@class='detailsStats']/b"));
    $this->assertEquals($numOfViews+1, $numOfViewsAfter);

    //test the download counter
    $numOfDownloads = intval($this->selenium->getText("xpath=//p[@class='detailsStats'][2]/b"));
    $this->selenium->click("xpath=//div[@class='detailsDownloadButton']/a[1]");
    $this->selenium->waitForPageToLoad(2000);
    $this->selenium->keyPressNative("27"); //press escape key
    $this->selenium->refresh();
    $this->selenium->waitForPageToLoad(10000);
    $numOfDownloadsAfter = intval($this->selenium->getText("xpath=//p[@class='detailsStats'][2]/b"));
    $this->assertEquals($numOfDownloads+1, $numOfDownloadsAfter);
    $this->selenium->click("xpath=//div[@class='detailsMainImage']/a[1]");
    $this->selenium->waitForPageToLoad(2000);
    $this->selenium->keyPressNative("27"); //press escape key
    $this->selenium->refresh();
    $this->selenium->waitForPageToLoad(10000);
    $numOfDownloadsAfter = intval($this->selenium->getText("xpath=//p[@class='detailsStats'][2]/b"));
    $this->assertEquals($numOfDownloads+2, $numOfDownloadsAfter);

    //test the home link
    $this->selenium->click("xpath=//div[@class='webHeadTitleName']/a");
    $this->selenium->waitForPageToLoad(10000);
    $this->assertRegExp("/catroid\/index/", $this->selenium->getLocation());
    $this->selenium->goBack();
    $this->selenium->waitForPageToLoad(10000);
    $this->selenium->click("xpath=//div[@class='webHeadLogo']/a");
    $this->selenium->waitForPageToLoad(10000);
    $this->assertRegExp("/catroid\/index/", $this->selenium->getLocation());
    $this->selenium->goBack();
    $this->selenium->waitForPageToLoad(10000);
  }

  /**
   * @dataProvider randomIds
   */
  public function testInappropriateButton($id) {
    $this->loginAsAdministrator();
    $this->selenium->open(TESTS_BASE_PATH.'catroid/details/'.$id);
    $this->selenium->waitForPageToLoad(10000);
    $this->assertTrue($this->selenium->isElementPresent("reportAsInappropriateButton"));
    $this->selenium->click("reportAsInappropriateButton");
    $this->assertTrue($this->selenium->isVisible("reportInappropriateReason"));
    $this->assertTrue($this->selenium->isVisible("reportInappropriateReportButton"));
    $this->assertTrue($this->selenium->isVisible("reportInappropriateCancelButton"));
    $this->selenium->click("reportAsInappropriateButton");
    $this->assertFalse($this->selenium->isVisible("reportInappropriateReason"));
    $this->selenium->click("reportAsInappropriateButton");
    $this->assertTrue($this->selenium->isVisible("reportInappropriateReason"));
    $this->selenium->click("reportInappropriateCancelButton");
    $this->assertFalse($this->selenium->isVisible("reportInappropriateReason"));
    $this->selenium->click("reportAsInappropriateButton");
    $this->selenium->click("reportInappropriateReportButton");
    $this->selenium->waitForPageToLoad(10000);
    $this->assertFalse($this->selenium->isVisible("reportInappropriateReason"));
    $this->assertFalse($this->selenium->isTextPresent("You reported this project as inappropriate!"));
    $this->selenium->click("reportAsInappropriateButton");
    $this->selenium->type("reportInappropriateReason", "my selenium reason");
    $this->selenium->click("reportInappropriateReportButton");
    $this->selenium->waitForPageToLoad(10000);
    $this->assertFalse($this->selenium->isVisible("reportInappropriateReason"));
    $this->assertTrue($this->selenium->isTextPresent("You reported this project as inappropriate!"));
    if(TESTS_BROWSER != "*googlechrome") { //chrome does not support keyPress for some reason
	    $this->selenium->refresh();
	    $this->selenium->waitForPageToLoad(10000);
	    $this->selenium->click("reportAsInappropriateButton");
	    $this->selenium->type("reportInappropriateReason", "my selenium reason 2");
	    $this->selenium->focus("reportInappropriateReason");
	    $this->selenium->keyPress("reportInappropriateReason", "\\13");
	    $this->selenium->waitForPageToLoad(10000);
	    $this->assertFalse($this->selenium->isVisible("reportInappropriateReason"));
	    $this->assertTrue($this->selenium->isTextPresent("You reported this project as inappropriate!"));
    }
    //unflag the project again
    $path= 'http://'.ADMIN_AREA_USER.':'.DB_PASS.'@'.str_replace('http://', '', TESTS_BASE_PATH).'admin/tools/inappropriateProjects';
    $this->selenium->open($path);
    $this->selenium->waitForPageToLoad(10000);
    $this->assertTrue($this->selenium->isTextPresent($id));
    $this->selenium->click("resolve".$id);
    $this->selenium->waitForPageToLoad(10000);
    $this->assertTrue($this->selenium->isTextPresent("The project was succesfully restored and set to visible!"));
    $this->assertFalse($this->selenium->isTextPresent($id));
  }

  /**
   * @dataProvider titlesAndDescriptions
   */
  public function testMoreButtonAndQRCode($title, $description) {
    $this->uploadTestProject($title, $description);

    $projectTitle = $title;
    $projectDescription = $description;

    //test more button
    $this->selenium->open(TESTS_BASE_PATH);
    $this->selenium->waitForPageToLoad("10000");
    $this->ajaxWait();
    $this->selenium->click("xpath=//a[@class='projectListDetailsLink']");
    $this->selenium->waitForPageToLoad("10000");
    // $this->ajaxWait("class=showFullDescriptionButton");

    $this->assertTrue($this->selenium->isElementPresent("showFullDescriptionButton"));
    $shortDescriptionFromPage = $this->selenium->getText("xpath=//p[@id='detailsDescription']");
    $this->assertNotEquals(trim($shortDescriptionFromPage), trim($projectDescription));
    $this->selenium->click("showFullDescriptionButton");
    $fullDescriptionFromPage = $this->selenium->getText("xpath=//p[@id='detailsDescription']");
    $this->assertEquals(trim($fullDescriptionFromPage), trim($projectDescription));
    $this->assertNotEquals(trim($fullDescriptionFromPage), trim($shortDescriptionFromPage));
    $this->assertTrue($this->selenium->isElementPresent("showShortDescriptionButton"));
    $this->selenium->click("showShortDescriptionButton");
    $this->assertEquals($shortDescriptionFromPage, $this->selenium->getText("xpath=//p[@id='detailsDescription']"));

    //test qr code
    $lastProject = $this->getLastProject();
    $projectId = $lastProject['id'];
    $qrCodeFile = CORE_BASE_PATH.PROJECTS_QR_DIRECTORY.$projectId.PROJECTS_QR_EXTENTION;
    $qrCodeFileNew = CORE_BASE_PATH.PROJECTS_QR_DIRECTORY.$projectId.'new'.PROJECTS_QR_EXTENTION;
    if(file_exists($qrCodeFile)) {
      $this->assertTrue($this->selenium->isElementPresent("xpath=//img[@class='projectDetailsQRImage']"));
      rename($qrCodeFile, $qrCodeFileNew);
      $this->selenium->refresh();
      $this->selenium->waitForPageToLoad("10000");
      $this->assertFalse($this->selenium->isElementPresent("xpath=//img[@class='projectDetailsQRImage']"));
      rename($qrCodeFileNew, $qrCodeFile);
      $this->selenium->refresh();
      $this->selenium->waitForPageToLoad("10000");
      $this->assertTrue($this->selenium->isElementPresent("xpath=//img[@class='projectDetailsQRImage']"));
    } else {
      $this->assertFalse($this->selenium->isElementPresent("xpath=//img[@class='projectDetailsQRImage']"));
    }

    // Test QR-Code Info
    $this->assertTrue($this->selenium->isElementPresent("xpath=//button[@id='showQrCodeInfoButton']"));
    $this->assertTrue($this->selenium->isElementPresent("xpath=//div[@id='qrcodeInfo']"));
    $this->assertTrue($this->selenium->isElementPresent("xpath=//button[@id='hideQrCodeInfoButton']"));
    $this->assertTrue($this->selenium->isVisible("xpath=//button[@id='showQrCodeInfoButton']"));
    $this->assertFalse($this->selenium->isVisible("xpath=//button[@id='hideQrCodeInfoButton']"));
    $this->assertFalse($this->selenium->isVisible("xpath=//div[@id='qrcodeInfo']"));
    $this->selenium->click("showQrCodeInfoButton");
    $this->assertFalse($this->selenium->isVisible("xpath=//button[@id='showQrCodeInfoButton']"));
    $this->assertTrue($this->selenium->isVisible("xpath=//button[@id='hideQrCodeInfoButton']"));
    $this->assertTrue($this->selenium->isVisible("xpath=//div[@id='qrcodeInfo']"));
    $this->selenium->click("hideQrCodeInfoButton");
    $this->assertTrue($this->selenium->isVisible("xpath=//button[@id='showQrCodeInfoButton']"));
    $this->assertFalse($this->selenium->isVisible("xpath=//button[@id='hideQrCodeInfoButton']"));
    $this->assertFalse($this->selenium->isVisible("xpath=//div[@id='qrcodeInfo']"));
    
    //delete the test project
    $adminpath = 'http://'.ADMIN_AREA_USER.':'.DB_PASS.'@'.str_replace('http://', '', TESTS_BASE_PATH).'admin';
    $this->selenium->open($adminpath);
    $this->selenium->waitForPageToLoad(10000);
    $this->selenium->click("aAdministrationTools");
    $this->selenium->waitForPageToLoad(10000);
    $this->selenium->click("aAdminToolsEditProjects");
    $this->selenium->waitForPageToLoad(10000);
    $this->selenium->chooseOkOnNextConfirmation();
    $this->selenium->click("xpath=//input[@name='deleteButton']");
    $this->selenium->waitForPageToLoad(10000);
  }

  /* *** DATA PROVIDERS *** */
  //choose random ids from database
  public function randomIds() {
    $returnArray = array();

    $query = 'SELECT * FROM projects WHERE visible=true ORDER BY random() LIMIT 1';
    $result = pg_query($query) or die('DB operation failed: ' . pg_last_error());
    $projects = pg_fetch_all($result);
    pg_free_result($result);
    $i=0;
    foreach($projects as $project) {
      $returnArray[$i] = array($project['id'], $project['title'], $project['description']);
      $i++;
    }

    return $returnArray;
  }

  public function titlesAndDescriptions() {
    $returnArray = array(
    array('more button selenium test', 'This is a description which should have more characters than defined by the threshold in config.php. And once again: This is a description which should have more characters than defined by the threshold in config.php. Thats it!'),
    array('more button special chars test', utf8_decode('This is a description which has special chars like ", \' or < and > in it and it should have more characters than defined by the threshold in config.php. And once again: This is a description with "special chars" and should have more characters than defined by the threshold in config.php. Thats it!'))
    );

    return $returnArray;
  }

  private function getLastProject() {
    $query = 'SELECT * FROM projects ORDER BY upload_time DESC LIMIT 1';
    $result = pg_query($query) or die('DB operation failed: ' . pg_last_error());
    $projects = pg_fetch_all($result);
    pg_free_result($result);
    if(count($projects)) {
      return $projects[0];
    } else {
      return false;
    }
  }

  private function loginAsAdministrator() {
    $adminpath = 'http://'.ADMIN_AREA_USER.':'.DB_PASS.'@'.str_replace('http://', '', TESTS_BASE_PATH).'admin';
    $this->selenium->open($adminpath);
    $this->selenium->waitForPageToLoad(10000);
  }

  // upload a test project via cURL-request
  private function uploadTestProject($title, $description = '')
  {
    $uploadTestFile = dirname(__FILE__);
    if(strpos($uploadTestFile, '\\')  != false) {
      $uploadTestFile.= '\testdata\test.zip';
    } else {
      $uploadTestFile.= '/testdata/test.zip';
    }
    $uploadpath= TESTS_BASE_PATH.'catroid/upload/upload.json';

    $ch = curl_init();
    curl_setopt($ch, CURLOPT_HEADER, 0);
    curl_setopt($ch, CURLOPT_VERBOSE, 0);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_USERAGENT, "Mozilla/4.0 (compatible;)");
    curl_setopt($ch, CURLOPT_URL, $uploadpath);
    curl_setopt($ch, CURLOPT_POST, true);
    $post = array(
        "upload"=>"@$uploadTestFile",
        "projectTitle"=>$title,
        "projectDescription"=>$description,
        "fileChecksum"=>md5_file($uploadTestFile)
    );
    curl_setopt($ch, CURLOPT_POSTFIELDS, $post);
    $response = json_decode(curl_exec($ch));
    $this->assertEquals(200, $response->statusCode);
  }
}
?>

File: tests/selenium-grid/src/at/tugraz/ist/catroweb/catroid/DetailsTests.java
/**
  *Catroid: An on-device visual programming system for Android devices
  *Copyright (C) 2010-2013 The Catrobat Team
  *(<http://developer.catrobat.org/credits>)
  *
  *This program is free software: you can redistribute it and/or modify
  *it under the terms of the GNU Affero General Public License as
  *published by the Free Software Foundation, either version 3 of the
  *License, or (at your option) any later version.
  *
  *An additional term exception under section 7 of the GNU Affero
  *General Public License, version 3, is available at
  *http://developer.catrobat.org/license_additional_term
  *
  *This program is distributed in the hope that it will be useful,
  *but WITHOUT ANY WARRANTY; without even the implied warranty of
  *MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  *GNU Affero General Public License for more details.
  *
  *You should have received a copy of the GNU Affero General Public License
  *along with this program. If not, see <http://www.gnu.org/licenses/>.
  */

package at.tugraz.ist.catroweb.catroid;

import java.util.HashMap;

import org.openqa.selenium.By;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import static org.testng.AssertJUnit.*;

import at.tugraz.ist.catroweb.BaseTest;
import at.tugraz.ist.catroweb.common.*;

@Test(groups = { "catroid", "DetailsTests" })
public class DetailsTests extends BaseTest {

  @Test(dataProvider = "detailsProject", groups = { "functionality", "upload" }, description = "view + download counter test")
  public void detailsPageCounterLink(HashMap<String, String> dataset) throws Throwable {
    try {
      String response = projectUploader.upload(dataset);
      String id = CommonFunctions.getValueFromJSONobject(response, "projectId");
      String title = dataset.get("projectTitle");
      int numOfViews = -1;
      int numOfViewsAfter = -1;
      int numOfDownloads = -1;
      int numOfDownloadsAfter = -1;

      By viewsElement = By.xpath("//*[@id='projectDetailsContainer']/div[5]/ul/li[5]/div[2]");
      By downloadsElement = By.xpath("//*[@id='projectDetailsContainer']/div[5]/ul/li[4]/div[2]/span");
      By filesizeElement = By.xpath("//*[@id='projectDetailsContainer']/div[5]/ul/li[3]/div[2]");
      By downloadsButton = By.xpath("//*[@id='projectDetailsContainer']/div[3]/div/a[1]/div/span");

      openLocation("details/" + id);
      ajaxWait();
      // project title
      assertTrue(containsElementText(By.id("projectDetailsProjectTitle"), title.toUpperCase()));
      // test the view counter
      numOfViews = Integer.parseInt(driver().findElement(viewsElement).getText().split(" ")[0]);

      driver().navigate().refresh();
      ajaxWait();
      numOfViewsAfter = Integer.parseInt(driver().findElement(viewsElement).getText().split(" ")[0]);
      assertEquals(numOfViews + 1, numOfViewsAfter);

      // test the download counter
      numOfDownloads = Integer.parseInt(driver().findElement(downloadsElement).getText().split(" ")[0]);
      driver().findElement(downloadsButton).click();

      driver().navigate().refresh();
      ajaxWait();
      numOfDownloadsAfter = Integer.parseInt(driver().findElement(downloadsElement).getText().split(" ")[0]);
      assertEquals(numOfDownloads + 1, numOfDownloadsAfter);

      // check file size
      double filesize = CommonFunctions.getFileSizeRounded(Config.FILESYSTEM_BASE_PATH + Config.PROJECTS_DIRECTORY + id + Config.PROJECTS_EXTENTION);
      String filesizeText = driver().findElement(filesizeElement).getText();
      String displayedfilesize = filesizeText.substring(0, filesizeText.indexOf(" MB"));
      if(displayedfilesize.startsWith("<")) {
        // smaller files are displayed as "< 0.1 MB"
        assertEquals("< " + String.valueOf(filesize), displayedfilesize);
      } else {
        assertEquals(String.valueOf(filesize), displayedfilesize);
      }

      HashMap<String, String> versionInfo = CommonFunctions.getVersionInfo(id);
      String versionInfoText = driver().findElement(By.id("projectDetailsDownloadVersion")).getText();
      assertEquals("Pocket Code version: " + versionInfo.get("version_name"), versionInfoText);
    } catch(AssertionError e) {
      captureScreen("DetailsTests.detailsPageCounterLink." + dataset.get("projectTitle"));
      throw e;
    } catch(Exception e) {
      captureScreen("DetailsTests.detailsPageCounterLink." + dataset.get("projectTitle"));
      throw e;
    }
  }

  @Test(dataProvider = "detailsProject", groups = { "functionality", "upload" }, description = "view + download counter test")
  public void detailsPageCounterThumbnail(HashMap<String, String> dataset) throws Throwable {
    try {
      String response = projectUploader.upload(dataset);
      String id = CommonFunctions.getValueFromJSONobject(response, "projectId");
      int numOfDownloads = -1;
      int numOfDownloadsAfter = -1;
      
      By downloadsElement = By.xpath("//*[@id='projectDetailsContainer']/div[5]/ul/li[4]/div[2]/span");
      
      openLocation("details/" + id);
      
      // test the download counter
      numOfDownloads = Integer.parseInt(driver().findElement(downloadsElement).getText().split(" ")[0]);
      driver().findElement(By.id("projectDetailsThumbnailImage")).click();
      driver().navigate().refresh();
      ajaxWait();
      numOfDownloadsAfter = Integer.parseInt(driver().findElement(downloadsElement).getText().split(" ")[0]);
      assertEquals(numOfDownloads + 1, numOfDownloadsAfter);
    } catch(AssertionError e) {
      captureScreen("DetailsTests.detailsPageCounterThumbnail." + dataset.get("projectTitle"));
      throw e;
    } catch(Exception e) {
      captureScreen("DetailsTests.detailsPageCounterThumbnail." + dataset.get("projectTitle"));
      throw e;
    }
  }

  @Test(dataProvider = "detailsProject", groups = { "functionality", "upload" }, description = "test inappropriate button")
  public void inappropriateButton(HashMap<String, String> dataset) throws Throwable {
    try {
      String response = projectUploader.upload(dataset);
      String id = CommonFunctions.getValueFromJSONobject(response, "projectId");

      openLocation("details/" + id);
      assertTrue(isElementPresent(By.id("reportAsInappropriateButton")));
      
      driver().findElement(By.id("reportAsInappropriateButton")).click();
      ajaxWait();
      driver().findElement(By.xpath("//*[@id='reportAsInappropriateDialog']/a")).click();
      ajaxWait();

      driver().findElement(By.id("loginUsername")).sendKeys(CommonData.getLoginUserDefault());
      driver().findElement(By.id("loginPassword")).sendKeys(CommonData.getLoginPasswordDefault());
      driver().findElement(By.id("loginSubmitButton")).click();
      ajaxWait();

      assertTrue(isElementPresent(By.id("reportAsInappropriateButton")));
      driver().findElement(By.id("reportAsInappropriateButton")).click();
      assertTrue(isVisible(By.id("reportInappropriateReason")));
      assertTrue(isVisible(By.id("reportInappropriateReportButton")));
      assertTrue(isVisible(By.id("reportInappropriateCancelButton")));
      driver().findElement(By.id("reportAsInappropriateButton")).click();
      assertFalse(isVisible(By.id("reportInappropriateReason")));
      driver().findElement(By.id("reportAsInappropriateButton")).click();
      assertTrue(isVisible(By.id("reportInappropriateReason")));
      driver().findElement(By.id("reportInappropriateCancelButton")).click();
      assertFalse(isVisible(By.id("reportInappropriateReason")));
      driver().findElement(By.id("reportAsInappropriateButton")).click();
      driver().findElement(By.id("reportInappropriateReportButton")).click();
      ajaxWait();
      assertFalse(isVisible(By.id("reportInappropriateReason")));
      assertFalse(isTextPresent("You reported this project as inappropriate!"));
      driver().findElement(By.id("reportAsInappropriateButton")).click();
      driver().findElement(By.id("reportInappropriateReason")).sendKeys("my selenium reason");
      driver().findElement(By.id("reportInappropriateReportButton")).click();
      ajaxWait();
      assertFalse(isVisible(By.id("reportInappropriateReason")));
      ajaxWait();
      assertTrue(isTextPresent("You reported this project as inappropriate!"));

      driver().navigate().refresh();
      ajaxWait();
      assertTrue(isTextPresent("No entry was found for the given ID"));

      openAdminLocation("/tools/inappropriateProjects");
      clickOkOnNextConfirmationBox();
      driver().findElement(By.id("resolve" + id)).click();
      assertTrue(isTextPresent("The project was succesfully restored and set to visible!"));
      assertFalse(isTextPresent(id));
    } catch(AssertionError e) {
      captureScreen("DetailsTests.inappropriateButton." + dataset.get("projectTitle"));
      throw e;
    } catch(Exception e) {
      captureScreen("DetailsTests.inappropriateButton." + dataset.get("projectTitle"));
      throw e;
    }
  }

  @Test(groups = { "visibility" }, description = "check if invalid project id redirects to errorpage")
  public void invalidProjectID() throws Throwable {
    try {
      String invalidProject = CommonData.getRandomShortString(10);
      openLocation("details/" + invalidProject);
      assertRegExp(".*/error", driver().getCurrentUrl());
      assertTrue(isTextPresent("No entry was found for the given ID:"));
      assertTrue(isTextPresent("ID: " + invalidProject));
      assertFalse(isElementPresent(By.xpath("//div[@class='detailsFlexDiv']")));
    } catch(AssertionError e) {
      captureScreen("DetailsTests.invalidProjectID");
      throw e;
    } catch(Exception e) {
      captureScreen("DetailsTests.invalidProjectID");
      throw e;
    }
  }

  @DataProvider(name = "titlesAndDescriptions")
  public Object[][] titlesAndDescriptions() {
    Object[][] returnArray = new Object[][] {
        { CommonData
          .getUploadPayload(
              "more button selenium test",
              "This is a description which should have more characters than defined by the threshold in config.php. And once again: This is a description which should have more characters than defined by the threshold in config.php. Thats it!",
              "", "", "", "", "", "") },
              { CommonData
                .getUploadPayload(
                    "more button special chars test",
                    "This is a description which has special chars like \", & or < and > in it and it should have more characters than defined by the threshold in config.php. And once again: This is a description with \"special chars\" and should have more characters than defined by the threshold in config.php. Thats it!",
                    "", "", "", "", "", "") }, };
    return returnArray;
  }

  @DataProvider(name = "detailsProject")
  public Object[][] detailsProject() {
    Object[][] returnArray = new Object[][] {
        { CommonData.getUploadPayload("details_test1small", "details_test_description", "", "", "", "", "catroid", CommonFunctions.getAuthenticationToken("catroid")) },
        { CommonData.getUploadPayload("details_test2big", "details_test_description", "", "", "", "", "catroid", CommonFunctions.getAuthenticationToken("catroid")) }, };
    return returnArray;
  }
}